### ISSUE TITLE
Spacing after keywords

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/syn#17]
Reported by email from @gregkatz:

> I ran into a little bit of a problem doing the syn parsers. It's a little more complicated than I realized. Specifically, I was experimenting with the expression after the while keyword to see what works and what doesn't in the actual Rust compiler, and my parser doesn't actually match the behavior of the compiler. For example here's how the compiler handles the following:

> ```
whilevariable {} //error
whiletrue //error
while{variable} {} //ok
while(variable) {} //ok
while1<2 {} //error
while-1<2 {} //ok
while&1<&2 {} //ok
while*variable {} //ok
while!variable {} //ok
```

> I believe my parser as it currently stands would allow all of this, but I'm not sure how to make it match the behavior of the real compiler.

**Comment by dtolnay** [comment]
I have a fix in #16. This is caused by abusing `punct!(...)` (i.e. "punctuation") for parsing keywords. I added a separate `keyword!(...)` parser and updated all the existing calls.

**Comment by dtolnay** [comment]
And one more... 57ffbde59d3c0b9cd051ad1a3a54765f986942fa
